File: Session12/22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        temp=[]
        ans=[]
        def helper(maxparan,open,close):
            if len(temp)>=2*maxparan:
                ans.append("".join(temp))
                return
            if open<maxparan:
                temp.append("(")
                helper(maxparan,open+1,close)
                temp.pop(-1)
            if close<open:
                temp.append(")")
                helper(maxparan,open,close+1)
                temp.pop(-1)
        
        helper(n,0,0)
        return ans

File: Session12/22. Generate Parentheses.java
class Solution {
    List<String> ans=new ArrayList();
    public List<String> generateParenthesis(int n) {
        backtrack(new StringBuilder(),0,0,n);
        return ans;
    }
    public void backtrack(StringBuilder curr,int open,int close,int max)
    {
        if(curr.length()==2*max)
        {
            ans.add(curr.toString());
        }
        if(open<max)
        {
            curr.append("(");
            backtrack(curr,open+1,close,max);
            curr.setLength(curr.length()-1);
        }
        if(close<open)
        {
            curr.append(")");
            backtrack(curr,open,close+1,max);
            curr.setLength(curr.length()-1);
        }
    }
}